Documentation

U8NWXD opened the pull request:
**Deadline:** September 18th
* Remove outdated material and add disclaimers for missing sections

---------

Docs are currently being built on readthedocs: https://vivarium-core.readthedocs.io/en/docs/. This is the version of the docs from the `docs` branch, but `latest` and `stable` will update once we merge this PR.

Summary:

* Add the hand-written documentation from the old Vivarium repo
* Auto-generate docs from docstrings
* Move example glucose phosphorylation process, compartment, and experiment into vivarium-core from vivarium-cell since they're used in the docs for demonstration
* Move injector from vivarium-cell into vivarium-core since it is generally useful and used in the docs
* Fix references and syntax throughout the docstrings and written docs
* Update release script to clear `dist` directory

TODO:

* [x] Re-write docs to reflect recent changes
* [x] Add tests back and run in CI
* [ ] Give @eagmon and @prismofeverything access on readthedocs
* [x] Remove the code moved in from vivarium-cell from vivarium-cell
* [x] Make sure readthedocs webhook is working
* [x] Update getting started guide to remove zookeeper and kafka (and rest of docs)
* [x] Update docs link in README
* [ ] Update description of modeling paradigm
* [x] Write docs for package installation
* [x] Create template repo

eagmon commented:
@U8NWXD -- thanks for getting this migrated and updated!

eagmon commented:
@U8NWXD -- do you have any thoughts about how to combine process references from other libraries into this documentation?  For example, to https://vivarium-core.readthedocs.io/en/docs/reference/api/vivarium.processes.html.  ```vivarium-cell``` has many processes with good documentation.

U8NWXD commented:
@eagmon I was thinking of having two documentation sites: one for vivarium-core and one for vivarium-cell. Vivarium-cell would have the documentation on the individual processes.

I'm trying to avoid referencing vivarium-cell processes in vivarium-core to keep the vivarium-core docs self-contained. I'll have links in vivarium-cell to vivarium-core, but not the other way around, to avoid circular references. The idea here is that vivarium-core should not depend in any way on how other people use it, e.g. in vivarium-cell, including documentation. In this PR I created some basic processes, compartments, and experiments in vivarium-core so we can demonstrate how vivarium-core can be used without making people look at vivarium-cell. Does this make sense?

U8NWXD commented:
@eagmon This PR is almost ready. I've removed the old references to zookeeper and Kafka, and I updated the manually written docs to reflect our recent changes. One thing left is that the [description of Vivarium's modeling paradigm](https://vivarium-core.readthedocs.io/en/docs/guides/modeling_paradigm.html) still focuses on agent-based and whole-cell modeling. We should think about whether this should change

U8NWXD commented:
Another note: The docs are written around people cloning the GitHub repo still instead of installing the pip package. This is helpful because the repo has a demo process, compartment, and experiment users can experiment with, but we don't actually want people to use the code this way. We could create a separate demo repository for users to get started with? This could be that template repo we discussed at one point.

eagmon commented on AUTHORS.md:
```diff
@@ -0,0 +1,11 @@
+  * [Eran Agmon](https://eagmon.tumblr.com/home): `eagmon AT stanford DOT edu`
+  * Ryan Spangler: `ryan.spangler AT gmail DOT com`
+  * [Chris Skalnik](https://u8nwxd.github.io)
+  * Jerry Morrison: `j.erry.morrison AT gmail DOT com`
```
jerry.morrison

U8NWXD replied:
This is actually how Jerry asked that his email be specified. Gmail ignores dots when resolving email addresses, so he'll still get emails to this address. However, if a spammer gets this address, Jerry can tell Gmail to delete emails to `j.erry.morrison@...` to block the spam. As a side note, Gmail also ignores everything after `+` but before `@` so you can do things like this: `example+<anything you like>@gmail.com`

eagmon commented on doc/_static/draw_tree.py:
```diff
@@ -0,0 +1,35 @@
+        font_size=14,
+        ax=ax
+    )
+    plt.savefig('tree.png')
```
Looks intriguing! I didn't notice this before. Is this displayed in the docs?

U8NWXD replied:
Yeah, it's in the [hierarchy guide](https://vivarium-core.readthedocs.io/en/docs/guides/hierarchy.html). The figure needs some aesthetic work though

eagmon commented on doc/getting_started.rst:
```diff
@@ -0,0 +1,277 @@
+.. todo:: Use ``py -m agent.boot --host ip.to.remote.cluster:9092``
+    for remote Kafka services
+
+GNU Linear Programming Kit (GLPK)
```
I believe this is only needed for ```vivarium-cell```... Points to the issue of having multiple possible ```getting_started``` docs for different repos.

U8NWXD replied:
removed in f1f742fff55012cf7398b512f3ef1928fbdfbc14

eagmon commented on doc/guides/compartments.rst:
```diff
@@ -0,0 +1,295 @@
+    {
+        'sodium_pump': {
+            'cytoplasm': 'cell',
+            'extracellularSpace': 'bloodVessel',
```
we now link to stores in tuples to specify their relative compartment: ```('cell',)```. Might be worth giving an example of going up one level: ```('..', 'boundary')```

Edit -- you describe this below already. That looks sufficient.

U8NWXD replied:
I'll fix this part to use paths to be consistent. I also describe paths in more detail in the hierarchy guide

eagmon commented on doc/guides/compartments.rst:
```diff
@@ -0,0 +1,295 @@
+node. This tree is analogous to directory trees on a filesystem, and we
+use tuples of store names to specify a path through this tree. We call
+this tree the hierarchy, and we discuss it in more detail in the
+:doc:`hierarhcy guide <hierarchy>`.
```
typo - "hierarchy guide"

U8NWXD replied:
Fixed in 7605b0d

eagmon commented on doc/guides/hierarchy.rst:
```diff
@@ -0,0 +1,171 @@
+with a value that is a dictionary of :term:`schema keys`.
+
+---------------
+Hierarhcy Paths
```
typo -- "Hierarchy"

U8NWXD replied:
Fixed in 7605b0d

eagmon reviewed (approved):
Thanks @U8NWXD -- this looks great. Thank you for noting the needed update on ```modeling_paradigm```, I can take care of that in the next week or two.

I definitely think we should make a template repo. How's the name ```vivarium-template```? How can we coordinate the docs for that template and the core engine? Do you know of any other project that has the multiple library coordination challenge as we do?

Just a couple requests before you merge: Can you build this, add a link to the main core ```README```, and add me and Ryan to readthedocs.  Do we also have web hooks in place to automatically build the docs, like we had in the old vivarium?

U8NWXD commented:
Regarding template repo docs: My suggestion would be to not have docs for the template repo. That README should point to the vivarium-core docs. The vivarium-core docs can then link to the files in the template repo on GitHub instead of linking to their auto-generated docs. I don't see much point in having auto-generated docs for the template repo since the demo files aren't meant to be used in any models.

U8NWXD commented:
> Just a couple requests before you merge: Can you build this, add a link to the main core README, and add me and Ryan to readthedocs. Do we also have web hooks in place to automatically build the docs, like we had in the old vivarium?

Read the Docs is complaining that the webhook isn't working, but the documentation is building automatically like it should, so I'm ignoring the warning. For example, you can see that the typo in the hierarchy guide is fixed: [link](https://vivarium-core.readthedocs.io/en/docs/guides/hierarchy.html#hierarchy-paths).

This PR also changes the README to point to the updated docs. Once we merge the PR, Read the Docs will build the documentation for `master` and the link will point to those built docs.

I've added you as a maintainer on Read the Docs. @prismofeverything could you send me your username for Read the Docs so I can add you too?

U8NWXD commented on setup.py:
```diff
@@ -21,6 +21,10 @@
     license='MIT',
     entry_points={
         'console_scripts': []},
+    short_description=(
+        'Engine for combining computational biology models with the '
+        'Vivarium interface.'
+    ),
```
@eagmon does this description look good to you? If so, we can make the GitHub repo description match for consistency

eagmon replied:
How about this slightly modified description: `Engine for composing and simulating computational biology models with the Vivarium interface`. Yes to consistency with the repo description!

U8NWXD replied:
Done in 0bdd01e. I've also updated the repo description

U8NWXD commented:
@prismofeverything @eagmon FYI I added the tests I run on the documentation to GitHub actions now, so any changes that break the docs will be flagged on PRs. There are 2 tests I'm running:

1. A linter to check for typos and ensure the documentation writes phrases with multiple spellings consistently (e.g. timeseries vs time series vs time-series)
2. A test that builds the documentation and makes sure there are no errors. This ensures we don't have broken links, for example.

U8NWXD commented:
The template repository is up at https://github.com/vivarium-collective/vivarium-template, and I've re-written the getting started guide to use it

U8NWXD commented:
@eagmon I think this is ready to go in now. I'll wait until later today to merge though in case you notice something I need to fix

eagmon commented:
@U8NWXD -- this looks excellent, thank you! I'm excited about that template repo, we might already have a chance to use it today. Can you point me to those tests you mentioned?

U8NWXD commented:
@eagmon the template repo won't work quite yet. It needs vivarium-core to have some processes that are introduced by this PR

U8NWXD commented:
You can see the tests running here: https://github.com/vivarium-collective/vivarium-core/pull/6/checks?check_run_id=1129048015. They're defined here: https://github.com/vivarium-collective/vivarium-core/blob/docs/.github/workflows/docs.yml
